src/main/groovy/io/openliberty/tools/gradle/tasks/GenerateFeaturesTask.groovy

```diff
@@ -176,25 +176,31 @@ class GenerateFeaturesTask extends AbstractFeatureTask {
         def newServerXmlSrc = new File(server.configDirectory, GENERATED_FEATURES_FILE_PATH);
         try {
             if (missingLibertyFeatures.size() > 0) {
-                // Create specialized server.xml
-                ServerConfigXmlDocument configDocument = ServerConfigXmlDocument.newInstance();
-                configDocument.createComment(HEADER);
-                Element featureManagerElem = configDocument.createFeatureManager();
-                configDocument.createComment(featureManagerElem, GENERATED_FEATURES_COMMENT);
-                for (String missing : missingLibertyFeatures) {
-                    logger.debug(String.format("Adding missing feature %s to %s.", missing, GENERATED_FEATURES_FILE_PATH));
-                    configDocument.createFeature(missing);
+                Set<String> existingGeneratedFeatures = getGeneratedFeatures(servUtil, newServerXmlSrc);
+                if (!missingLibertyFeatures.equals(existingGeneratedFeatures)) {
+                    // Create specialized server.xml
+                    ServerConfigXmlDocument configDocument = ServerConfigXmlDocument.newInstance();
+                    configDocument.createComment(HEADER);
+                    Element featureManagerElem = configDocument.createFeatureManager();
+                    configDocument.createComment(featureManagerElem, GENERATED_FEATURES_COMMENT);
+                    for (String missing : missingLibertyFeatures) {
+                        logger.debug(String.format("Adding missing feature %s to %s.", missing, GENERATED_FEATURES_FILE_PATH));
+                        configDocument.createFeature(missing);
+                    }
+                    // Generate log message before writing file as the file change event kicks off other dev mode actions
+                    logger.lifecycle("Generated the following features: " + missingLibertyFeatures);
+                    // use logger.lifecycle so that message appears without --info tag on
+                    configDocument.writeXMLDocument(newServerXmlSrc);
+                    logger.debug("Created file " + newServerXmlSrc);
+                    // Add a reference to this new file in existing server.xml.
+                    def serverXml = findConfigFile("server.xml", server.serverXmlFile);
```

Comment:
I know this is existing behaviour, but will this reference still work if they have a custom server.xml file name configured in the build.gradle? It looks like it is searching in the source config folder?

Reply:
Tried this out and if I specify in the build.gradle a serverXmlFile location outside of the config directory, the generate features comment is not written. ie.
```gradle
liberty {
	server {
		serverXmlFile = file("src/main/liberty/testing/server1.xml")
	}
}
```
I realize we are doing the same in ci.maven as well: https://github.com/OpenLiberty/ci.maven/blob/339fc5ff2b312b3a29dafa632fa0a998b14f96fb/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/GenerateFeaturesMojo.java#L255

Maybe we should open a new issue to address this.

Reply:
Great, thank you.